dds_web/scheduled_tasks.py

```diff
@@ -1,24 +1,56 @@
 
 
-@scheduler.task("interval", id="available_to_expired", seconds=10, misfire_grace_time=1)
-def change_status_to_expired():
-    # print("Task: Change project status from Available to expired.", flush=True)
-    from dds_web.database import models
+@scheduler.task("cron", id="available_to_expired", minute=5, hour=2, misfire_grace_time=1)
+def set_available_to_expired():
+    scheduler.app.logger.debug("Task: Checking for Expiring projects.")
+    import sqlalchemy
 
-    # from dds_web.utils import current_time, page_query
+    from dds_web import db
+    from dds_web.database import models
+    from dds_web.errors import DatabaseError
+    from dds_web.api.project import ProjectStatus
+    from dds_web.utils import current_time, page_query
 
     with scheduler.app.app_context():
-        # scheduler.app.logger.debug("This means that the app context works!")
-        test = models.Project.query.all()
-        # print(test, flush=True)
 
-        # for project in page_query(
-        #     models.ProjectStatuses.query.filter(models.ProjectStatuses.deadline <= current_time())
-        # ):
+        for project in page_query(models.Project.query.filter(models.Project.is_active == 1)):
+
+            if "Available" in project.current_status and project.current_deadline <= current_time():
+                scheduler.app.logger.debug("Handling expiring project")
+                scheduler.app.logger.debug(
+                    "Project: %s has status %s and expires on: %s",
+                    project.id,
+                    project.current_status,
+                    project.current_deadline,
+                )
+                expire = ProjectStatus()
+                new_status_row = expire.expire_project(
+                    project=project, current_time=current_time(), deadline_in=30
+                )
+
+                project.project_statuses.append(new_status_row)
 
-        #     flask.current_app.logger.debug("Project: %s - Expires: %s", project, project.expires)
+                try:
+                    db.session.commit()
+                except (sqlalchemy.exc.OperationalError, sqlalchemy.exc.SQLAlchemyError) as err:
+                    flask.current_app.logger.exception(err)
+                    db.session.rollback()
+                    raise DatabaseError(
+                        message=str(err),
+                        alt_message=(
+                            "Status was not updated"
+                            + (
+                                ": Database malfunction."
+                                if isinstance(err, sqlalchemy.exc.OperationalError)
+                                else ": Server Error."
+                            )
+                        ),
+                    ) from err
+                scheduler.app.logger.debug("Project: %s has status Archived now!", project.title)
+            else:
+                scheduler.app.logger.debug("Nothing to do for Project: %s", project.title)
```

Comment:
Commiting once per project is fine but then you shouldn't raise the exception and instead save a list of the different errors that occurr and then log them after the scheduled task is done. So for example I would add a dict before the for loop(s) `errors = {}` and then in the unit loop `errors[unit.name] = {}`, and then in this except statement I'd do
```
errors[unit.name][project.public_id] = {"message": str(err), alt_message=("Status was not updated" + etc....)}
continue
```
And then after both for loops are finished 
```
for unit, projects in errors.items():
    scheduler.app.logger.error(f"Unit '{unit}' projects encountered issues during expiration process:")
    for project in projects:
        # log the project errors ("message" probably) for each project and note which unit it belongs to
        # or 
```

Reply:
What I noticed from testing with DB down is that the apscheduler is already catching the error when trying to access its own table apscheduler_jobs and raising an error 
`dds_backend     | Error getting due jobs from job store 'default': (pymysql.err.OperationalError) (2003, "Can't connect to MySQL server on 'db' ([Errno -2] Name or service not known)")`

This basically means that the jobs will not be triggered and there won't be errors coming from them. On the other hand DB can go down (or connection can get lost) after the jobs are triggered, which means that we still want to catch and log them.
I am thinking correctly here!

Reply:
I'm not entirely sure I'm following, but do you mean that if the commit for one project fails and therefore raises an exception, the for loop will still continue with the rest of the projects? This is what I meant with my comment, sorry if that wasn't clear. If there is a SQLAlchemyError when commiting one project status change, it could be that there's a problem with exactly that project and none of the others, so then the project status change should continue in the other ones.

Reply:
the changes are pushed now, not sure if I've done the error handling properly but let's see